[HUDI-3020] Utility to create manifest file

codejoyan opened the pull request:
Please refer https://issues.apache.org/jira/browse/HUDI-3020 for more details.

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

vingov commented:
@nsivabalan or @xushiyan - Can you please approve the workflow?

@codejoyan - Thanks for working on this feature, I tested the code it works well, I need you to remove the base path plus partition folder and just keep the filenames to make it work for the BigQuery sync, can you please update your code?

When you get a chance please add unit tests as well.

vinothchandar reviewed (commented):
Awesome! Can we add a simple unit test for this!

Can we move this to the `hudi-sync-common` module

codejoyan commented:
@nsivabalan or @xushiyan - Can you please approve the workflow?

I have made the below changes. @vingov please review when you get a chance. 
1. @vinothchandar  Moved the class to hudi-sync-common
2. Removed the base path plus partition folder and kept just the filenames
3. Added test cases

xushiyan commented on hudi-sync/hudi-sync-common/src/main/java/org/apache/hudi/sync/common/util/ManifestFileUtil.java:
```diff
@@ -0,0 +1,142 @@
+  public Stream<String> fetchLatestBaseFilesForAllPartitions() {
+    try {
+      HoodieMetadataConfig metadataConfig = buildMetadataConfig(hadoopConf.get());
+      HoodieMetadataFileSystemView fsView = new HoodieMetadataFileSystemView(engineContext, metaClient, metaClient.getActiveTimeline().getCommitsTimeline().filterCompletedInstants(), metadataConfig);
```
you should probably create a new HoodieMetadataFileSystemView within the lambda below, also create HoodieLocalEngineContext there instead of keep 1 for the entire object. HoodieLocalEngineContext is not serializeable.

codejoyan commented:
@xushiyan, Please review when you get a chance.

xushiyan commented on hudi-sync/hudi-sync-common/pom.xml:
```diff
@@ -104,6 +104,14 @@
       <scope>test</scope>
     </dependency>
+
+    <dependency>
+      <groupId>org.apache.hudi</groupId>
+      <artifactId>hudi-common</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
+      <scope>test</scope>
+    </dependency>
```
can you remove this ? hudi-common is added at the top already right?

xushiyan commented on hudi-sync/hudi-sync-common/src/test/java/org/apache/hudi/sync/common/util/TestManifestFileUtil.java:
```diff
@@ -0,0 +1,81 @@
+    String instant = "100";
+    hoodieTestTable = hoodieTestTable.addCommit(instant);
+    // Generate 10 files under each partition
+    MULTI_LEVEL_PARTITIONS.forEach(p -> {
+      try {
```
/nit prefer for-loop over foreach so you don't have to try-catch. just throw out looks cleaner

xushiyan commented on hudi-sync/hudi-sync-common/src/test/java/org/apache/hudi/sync/common/util/TestManifestFileUtil.java:
```diff
@@ -0,0 +1,81 @@
+    // Generate 10 files under each partition
+    createTestDataForPartitionedTable(10);
+    ManifestFileUtil mainfestFileUtil = ManifestFileUtil.builder().setConf(metaClient.getHadoopConf()).setBasePath(basePath).build();
+    try {
+      mainfestFileUtil.writeManifestFile();
+      Assertions.assertTrue(FSUtils.getFileSize(metaClient.getFs(), mainfestFileUtil.getManifestFilePath()) > 0);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
```
can you remove try-catch so any failure not silenced?

xushiyan reviewed (commented):
Almost there.

codejoyan commented:
@xushiyan Made changes in the test classes.

codejoyan commented:
Tested from spark-spell using the below steps:

val basePath = "gs://udp-hudi-storage5/hudi_demo_dataset"
val conf = new new Configuration()
val manifestFileUtil: ManifestFileUtil = ManifestFileUtil.builder().setConf(conf).setBasePath(basePath).build();
manifestFileUtil.writeManifestFile

bash-4.2$ gsutil cat gs://udp-hudi-storage5/hudi_demo_dataset/.hoodie/manifest/latest-snapshot.csv | head -5
95e78133-08dd-4721-9c49-8fbe338589f0-0_621-10-1021_20210927173651.parquet
1c01ddf1-41e8-43bf-94f3-9a75c9e39b21-0_1238-12-1638_20210927173651.parquet
a021864b-d0b6-462a-8f47-6370668993d6-0_694-12-1094_20210927173651.parquet
11815af9-6ed5-4f05-9b6b-7f79479f4f53-0_625-12-1025_20210927173651.parquet
ef4fc799-9124-4656-8106-f0374590dde6-0_1151-12-1551_20210927173651.parquet
